Lib/ftplib.py

```diff
@@ -432,16 +446,95 @@ def retrbinary(self, cmd, callback, blocksize=8192, rest=None):
+
         self.voidcmd('TYPE I')
-        with self.transfercmd(cmd, rest) as conn:
-            while 1:
-                data = conn.recv(blocksize)
-                if not data:
-                    break
-                callback(data)
-            # shutdown ssl layer
-            if _SSLSocket is not None and isinstance(conn, _SSLSocket):
-                conn.unwrap()
+        self.voidcmd('MODE %s' % self.transmissionmode)
+
+        if self.transmissionmode == 'S':
+            with self.transfercmd(cmd, rest) as conn:
+                while 1:
+                    data = conn.recv(blocksize)
+                    if not data:
+                        break
+                    callback(data)
+                # shutdown ssl layer
+                if _SSLSocket is not None and isinstance(conn, _SSLSocket):
+                    conn.unwrap()
+        elif self.transmissionmode == 'B':
+            with self.transfercmd(cmd, rest) as conn:
+                while 1:
+                    # Receive one byte at a time, not all 3 at once -- seriously
+                    header = bytes()
+                    for i in range(0, 3):
+                        header += conn.recv(1)
+
+                    (descriptor, blocklength) = struct.unpack('!BH', header)
+                    if self.debugging:
+                        print("*header* %d\t%d" % (descriptor, blocklength))
+
+                    if (0 <= descriptor <= 240) and descriptor % 16 == 0:
+                        pass
+                    else:
+                        # Abort the transfer. Expect 426 response code.
+                        self.abort()
+                        # Exception so close the data connection.
+                        self.close_dataconn()
+                        # Catch the 226 response code.
+                        self.voidresp()
+                        raise error_proto("Unexpected header descriptor. Data block is invalid.")
+
+                    data = bytes()
+                    dtl = []
+                    sread = 0
+                    while 1:
+                        if blocklength == 0:
+                            break
+                        try:
+                            buff = conn.recv(blocklength-sread, 0)
+                        except socket.error as se:
+                            if self.debugging:
+                                print("*got socket error: %s*" % se)
+                            raise se
+                        if not buff:
+                            if self.debugging:
+                                print("*got no data on recv* final size: %d" % sread)
+                            break
+                        if len(buff) > blocklength-sread:
+                            if self.debugging:
+                                print("*got more data than desired!* %d vs %d" % (len(buff), blocklength-sread))
+                        sread = sread + len(buff)
+                        data = data + buff
```

Comment:
Also, if the server sends us a big number (say size = 1G), we will hog RAM. It would probably makes sense to implement an additional, internal buffer, that calls `callback` function after a certain threshold (e.g. 1M) and empties the internal memory buffer.